### ISSUE TITLE
bin/setup-grunt dont work apple sillicon

### ISSUE THREAD

**Issue by luancschmitz** [markshust/docker-magento#569]
**Description**
on macs with apple sillicon, when running bin/setup-grunt an error occurs stating that phantonjs-prebuild is not available for the arm architecture.

**Steps To Reproduce**
1. run bin/setup-grunt in a store that uses the docker

**Expected Result**
1. The command runs without any error.

**Actual Result**
when running npm install(bin/setup-grunt) an error is returned stating that phantojs-prebuild is not available for arm.

**Comment by markshust** [comment]
I looked into this and did confirm the issue, just an FYI.

**Comment by sasafekete** [comment]
@luancschmitz @markshust If you remove the following line from package.json.sample, it works
"grunt-contrib-jasmine": "~1.1.0",

**Comment by luancschmitz** [comment]
Hi @sasafekete its works fine.

**Comment by markshust** [comment]
@sasafekete thanks for the update. I think we should hack the composer.json to update and remove jasmine then (or replace it with an alternative that works with apple silicon).

**Comment by markshust** [comment]
Here's what I used to do in `bin/setup` to eliminate the `vertex/module-tax` Composer module:

```
sed -e 's/"conflict": {/"replace": { "vertex\/module-tax": "*" },\ "conflict": {/' src/composer.json > composer.json && mv composer.json src/composer.json
```

Perhaps we should do something similar to this package.json.sample file.

**Comment by markshust** [comment]
I really, really tried to add PhantomJS or find a build that works with ARM chips, but I'm really deeming this unfixable. I'll proceed with adding in the above hack to `bin/setup-grunt`, because I also think not many even use Phantom anymore.

**Comment by luancschmitz** [comment]
> `npm ERR! code 1 npm ERR! path /var/www/html/node_modules/phantomjs-prebuilt npm ERR! command failed npm ERR! command sh -c node install.js npm ERR! PhantomJS not found on PATH npm ERR! Unexpected platform or architecture: linux/arm64 npm ERR! It seems there is no binary available for your platform/architecture npm ERR! Try to install PhantomJS globally
> 
> npm ERR! A complete log of this run can be found in: npm ERR! /var/www/.npm/_logs/2022-05-31T14_40_37_882Z-debug-0.log`
> 
> Someone can help me with this?

phantonjs-prebuild is not compatible with ARM64 (and I don't think it ever will be).
You can change the package,json.sample file and remove the line
"grunt-contrib-jasmine": "~1.2.0"

so bin/setup-grunt will work

**Comment by markshust** [comment]
Working on this

**Comment by markshust** [comment]
Hey everyone, pushed out some updates.

If an arm64 chip is detected, it will now automatically remove the `grunt-contrib-jasmine` module from the `package.json` file.

I also added `grunt-config.json`, `Gruntfile.js` and `package.json` files to the volume mounts in the `docker-compose.dev.yml` file. By default, these will be set to the `.sample` files that come with a vanilla install of Magento.

In the `bin/setup-grunt` file, I'm copying the `.sample` files over to the non-.sample filenames if they do not already exist, then updating the volume mounts to look at those files. This will allow you to modify these files and persist updates between Magento version updates.

I also fixed the bonked `bin/grunt` file as that was looking at the global `grunt` binary, which no longer exists. Now, it will look at the grunt binary at `./node_modules/grunt/bin/grunt`, which will now exist after running `bin/setup-grunt`.

Can y'all confirm it's now working after running `bin/update`?

**Comment by markshust** [comment]
Assuming all is working, so closing this one out. If you are still having issues, please let me know in a comment and I'll reopen this ticket.

**Comment by vncnt-mb** [comment]
Hello @markshust,
Thanks a lot for the fix, but, sorry to reopen this issue, I have a problem when running a grunt exec:$theme. I think the executable is still based on system grunt; here is the error:

```
bin/grunt exec:blank --force

Running "exec:blank" (exec) task
>> /bin/sh: 1: grunt: not found
>> Exited with code: 127.
>> Error executing child process: Error: Process exited with code 127.
Warning: Task "exec:blank" failed. Used --force, continuing.

Done, but with warnings.


Execution Time (2022-06-29 09:22:29 UTC-0)
```
to be sure, I retrieve the Gruntfile.js, package.json and grunt-config.json files from the Magento 2.4.4 directory; but no changes.

thanks in advance.

**Comment by vncnt-mb** [comment]
@markshust Yes, I agree with you, but since the change of the **bin/grunt** file to point in the **node_modules** directory, it's stop working :/

Grunt is well installed, but it seems lost.

I just tested **bin/cli npx grunt exec:luma** without installing **grunt-cli** globally and it works. So if we block the executable in the root folder of Magento, it would be perfect!

In my opinion, the best will be to indicate in the PATH environment variable that in the **./node_modules/bin** folder, there are several executable elements. Or, we can simply modify the **bin/grunt** file by integrating **bin/cli npx grunt "$@"**

If it can help to integrate a ready-to-use solution! :)

**Comment by schneider-vaimo** [comment]
@markshust I'm experiencing the same issue with the missing global `grunt-cli`.
Also the generated `local-theme.js` file was setting an empty string as my themes locale, instead of the expected `'nb_NO'`.
like so:
```
mytheme: {
    area: 'frontend',
    name: 'custom/mytheme',
    locale: '',
    files: [ 'css/styles-m', 'css/styles-l' ],
    dsl: 'less'
}
```
Also, is there any reason behind adding the magento themes to the `local-theme.js` file?

**Comment by kirjeldus** [comment]
I just want to confirm that I'm also having issues with bin/grunt on the latest php81-fpms (0 and 1) and I'm not on apple silicone and the fix suggested by vncnt-mb works well (just adding npx to the bin/grunt command).
 
> Also, is there any reason behind adding the magento themes to the `local-theme.js` file?

I have previously had some issues with variable inheritance, proper sourcemaps and live-reload and stuff like that if the default magento themes were not inlcuded in this file

**Comment by markshust** [comment]
Appreciate all the reports. I've been away from GitHub for a little while. I'm installing a base install of M2 with this project right now and will check this out.

**Comment by markshust** [comment]
Ok, I just added the above commit. I noticed that if you never ran the n98-magerun2 script, it wouldn't be installed the first time this script executed, which caused it to fail. So I added a safety check there.

I also used the npx fix above. I really have no clue what is going on and why the other line isn't working... but this works! So it's ok with me.

Here's how my `local-theme.js` file looks after the setup:

```
"use strict"; module.exports = {
  blank: {
    area: 'frontend',
    name: 'Magento/blank',
    locale: 'en_US',
    files: [ 'css/styles-m', 'css/styles-l', 'css/email', 'css/email-inline' ],
    dsl: 'less'
  },
  luma: {
    area: 'frontend',
    name: 'Magento/luma',
    locale: 'en_US',
    files: [ 'css/styles-m', 'css/styles-l' ],
    dsl: 'less'
  },
  backend: {
    area: 'adminhtml',
    name: 'Magento/backend',
    locale: 'en_US',
    files: [ 'css/styles-old', 'css/styles' ],
    dsl: 'less'
  }
}
```

Should this file be either copied over, and/or volume-synced back to the host after generation? Seems like this is a file that is likely to be modified over time and should be tracked in version control... no?

**Comment by luancschmitz** [comment]
Hi, this problem has fixe.
